src/adaptations/device-layer/EFR32/Logging.cpp

```diff
@@ -0,0 +1,500 @@
+#endif
+
+// How long to wait for non-error log messages to be dropped if spool queue is full. (Default to drop immediately)
+#ifndef LOG_TIMEOUT
+#define LOG_TIMEOUT (0)
+#endif
+
+// Maximum number of messages to queue to spooler
+#ifndef LOG_QUEUE_LEN
+#define LOG_QUEUE_LEN (50)
+#endif
+
+// Maximum size of each log message
+#ifndef LOG_ITEM_SZ
+#define LOG_ITEM_SZ WEAVE_DEVICE_CONFIG_LOG_MESSAGE_MAX_SIZE
+#endif
+
+#if WEAVE_DEVICE_CONFIG_ENABLE_THREAD
+#include <openthread/platform/logging.h>
+#endif // WEAVE_DEVICE_CONFIG_ENABLE_THREAD
+
+using namespace ::nl::Weave;
+using namespace ::nl::Weave::DeviceLayer;
+using namespace ::nl::Weave::DeviceLayer::Internal;
+
+static bool          sLogInitialized = false;
+static char          sMsgBuffer[LOG_ITEM_SZ];
+static QueueHandle_t sLogQueueHandle;
+static TaskHandle_t  sLogSpoolerTaskHandle;
+static uint8_t       sLogBuffer[LOG_RTT_BUFFER_SIZE];
+static bool          sHardFault = false;
+
+/**
+ * Enqueue a log message to the Logging Spooler task.  If
+ * this is called before tasks are started it prints directly to
+ * the RTT or UART.
+ */
+static void PrintLog(const char *msg, TickType_t timeout)
+{
+#if EFR32_LOG_ENABLED
+    if (sLogInitialized)
+    {
+        if (sHardFault == false && xTaskGetSchedulerState() == taskSCHEDULER_RUNNING)
+        {
+            xQueueSend(sLogQueueHandle, msg, timeout);
+        }
+        else
+        {
+#if RTT_LOGGING_ENABLED
+            size_t sz;
+            sz = strlen(msg);
+            SEGGER_RTT_WriteNoLock(0, msg, sz);
+            const char *newline = "\r\n";
+            sz                  = strlen(newline);
+            SEGGER_RTT_WriteNoLock(0, newline, sz);
+#endif
+        }
+    }
+#endif // EFR32_LOG_ENABLED
+}
+
+/**
+ * Logger task receives log messages from a queue and writes them to RTT or serial port via printf
+ */
+static void LogSpoolerTaskMain(void *pvParameter)
+{
+#if EFR32_LOG_ENABLED
+    while (1)
+    {
+        if (xQueueReceive(sLogQueueHandle, sMsgBuffer, portMAX_DELAY) == pdTRUE)
+        {
+#if RTT_LOGGING_ENABLED
+            size_t sz;
+            sz = strlen(sMsgBuffer);
+            SEGGER_RTT_WriteNoLock(0, sMsgBuffer, sz);
+            const char *newline = "\r\n";
+            sz                  = strlen(newline);
+            SEGGER_RTT_WriteNoLock(0, newline, sz);
+#endif
+        }
+    }
+#endif // EFR32_LOG_ENABLED
+}
+
+/**
+ * Initialize the serial port for logging
+ */
+extern "C" int efr32LogInit(void)
+{
+#if EFR32_LOG_ENABLED
+
+#if RTT_LOGGING_ENABLED
+    SEGGER_RTT_ConfigUpBuffer(LOG_RTT_BUFFER_INDEX, LOG_RTT_BUFFER_NAME, sLogBuffer, LOG_RTT_BUFFER_SIZE,
+                              SEGGER_RTT_MODE_NO_BLOCK_TRIM);
+#endif
+
+    sLogQueueHandle = xQueueCreate(LOG_QUEUE_LEN, LOG_ITEM_SZ);
+
+    if (sLogQueueHandle == NULL)
+    {
+        return -1;
+    }
+
+    if (xTaskCreate(LogSpoolerTaskMain, "logs", WEAVE_DEVICE_CONFIG_LOG_TASK_STACK_SIZE / sizeof(StackType_t), NULL,
```

Comment:
At some point I would like to eliminate the use of a separate task just for logging.  This seems like a lot of overhead as well as a potential source for log output latency.

Reply:
Removed separate logging task.  This was added initially as we were going to use the UART for logs instead of RTT.  Doing it this way made it easier for the the logging to be thread safe and to be able to drop logs.  Now reverted back to RTT.  We'd like add the option to use the UART for logging at a later date.

Reply:
Fully agree with adding support for UART logging, whenever we're ready.  Ideally this would be done without the need for a dedicated thread; e.g. by using an interrupt-based scheme.